src/main/java/picard/sam/markduplicates/ElcHashBasedDuplicatesFinder.java

```diff
@@ -58,92 +65,85 @@ void searchDuplicates(List<PairedReadSequence> sequences, Histogram<Integer> dup
         populateDupCandidates(sequences);
 
         //Duplicate set to filter the treated PairedReadSequences out
-        Set<PairedReadSequence> dupSet = new HashSet<>();
+        final Set<PairedReadSequence> dupSet = new HashSet<>();
+
         for (PairedReadSequence lhs : sequences) {
             if (dupSet.contains(lhs)) continue;
-
             final List<PairedReadSequence> dupes = new ArrayList<>();
-
-            for (PairedReadSequence rhs : lhs.dupCandidates) {
+            for (PairedReadSequence rhs : getSimilarReads(lhs)) {
                 if (dupSet.contains(rhs)) continue;
-
                 if (isDuplicate(lhs, rhs)) {
                     dupes.add(rhs);
                 }
             }
-
             dupSet.addAll(dupes);
             fillHistogram(duplicationHisto, opticalHisto, lhs, dupes);
         }
     }
 
+    private Set<PairedReadSequence> getSimilarReads(final PairedReadSequence pattern) {
+        final Set<PairedReadSequence> toCheck = new HashSet<>();
+        for (int[] hashesForRead: new int[][]{pattern.hashes1, pattern.hashes2}) {
+            for (int hash : hashesForRead) {
+                List<PairedReadSequence> readsWithSameHash = readsByHashInGroup.get(hash);
+                // if size == 1, then this list contains only pattern read
+                if (readsWithSameHash.size() > 1) {
+                    toCheck.addAll(readsWithSameHash);
+                }
+            }
+        }
+        return toCheck;
+    }
+
     /**
      * Populate PairedReadSequence.dupCandidates with duplicate candidates
      * based on the same hash on the same position.
      */
     private void populateDupCandidates(List<PairedReadSequence> seqs) {
         // Contains hash value as a key and PairedReadSequences match this key as a value
-        Map<Integer, List<PairedReadSequence>> readsByHash = new HashMap<>();
+        readsByHashInGroup.clear();
 
-        // Iterate over all PairedReadSequence and split in sets according to the hash value 
+        // Iterate over all PairedReadSequence and split in sets according to the hash value
         // in a certain position
         for (PairedReadSequence prs : seqs) {
             int[][] readHashValues = {prs.hashes1, prs.hashes2};
             for (int[] readHashValue : readHashValues) {
                 for (int key : readHashValue) {
-                    List<PairedReadSequence> dupCandidates = readsByHash.get(key);
-
-                    if (dupCandidates == null) {
-                        dupCandidates = new ArrayList<>();
-                        readsByHash.put(key, dupCandidates);
-                    }
+                    final List<PairedReadSequence> dupCandidates
```

Comment:
Actually, it's not cleaned because now it is in readsByHashInGroup map, and this list will contain all reads with identical hashes, it means that this reads will be checked for duplication.